README
NAME
    YAML::Syck - Fast, lightweight YAML loader and dumper

VERSION
    This document describes version 1.06 of YAML::Syck, released April 25,
    2009.

SYNOPSIS
        use YAML::Syck;

        # Set this for interoperability with other YAML/Syck bindings:
        # e.g. Load('Yes') becomes 1 and Load('No') becomes ''.
        $YAML::Syck::ImplicitTyping = 1;

        $data = Load($yaml);
        $yaml = Dump($data);

        # $file can be an IO object, or a filename
        $data = LoadFile($file);
        DumpFile($file, $data);

        # A string with multiple YAML streams in it
        $yaml = Dump(@data);
        @data = Load($yaml);

DESCRIPTION
    This module provides a Perl interface to the libsyck data serialization
    library. It exports the "Dump" and "Load" functions for converting Perl
    data structures to YAML strings, and the other way around.

    NOTE: If you are working with other language's YAML/Syck bindings (such
    as Ruby), please set $YAML::Syck::ImplicitTyping to 1 before calling the
    "Load"/"Dump" functions. The default setting is for preserving
    backward-compatibility with "YAML.pm".

FLAGS
  $YAML::Syck::Headless
    Defaults to false. Setting this to a true value will make "Dump" omit
    the leading "---\n" marker.

  $YAML::Syck::SortKeys
    Defaults to false. Setting this to a true value will make "Dump" sort
    hash keys.

  $YAML::Syck::SingleQuote
    Defaults to false. Setting this to a true value will make "Dump" always
    emit single quotes instead of bare strings.

  $YAML::Syck::ImplicitTyping
    Defaults to false. Setting this to a true value will make "Load"
    recognize various implicit types in YAML, such as unquoted "true",
    "false", as well as integers and floating-point numbers. Otherwise, only
    "~" is recognized to be "undef".

  $YAML::Syck::ImplicitUnicode
    Defaults to false. For Perl 5.8.0 or later, setting this to a true value
    will make "Load" set Unicode flag on for every string that contains
    valid UTF8 sequences, and make "Dump" return a unicode string.

    Regardless of this flag, Unicode strings are dumped verbatim without
    escaping; byte strings with high-bit set will be dumped with backslash
    escaping.

    However, because YAML does not distinguish between these two kinds of
    strings, so this flag will affect loading of both variants of strings.

  $YAML::Syck::ImplicitBinary
    Defaults to false. For Perl 5.8.0 or later, setting this to a true value
    will make "Dump" generate Base64-encoded "!!binary" data for all
    non-Unicode scalars containing high-bit bytes.

  $YAML::Syck::UseCode / $YAML::Syck::LoadCode / $YAML::Syck::DumpCode
    These flags control whether or not to try and eval/deparse perl source
    code; each of them defaults to false.

    Setting $YAML::Syck::UseCode to a true value is equivalent to setting
    both $YAML::Syck::LoadCode and $YAML::Syck::DumpCode to true.

BUGS
    Dumping Glob/IO values does not work yet.

CAVEATS
    This module implements the YAML 1.0 spec. To deal with data in YAML 1.1,
    please use the "YAML::XS" module instead.

    The current implementation bundles libsyck source code; if your system
    has a site-wide shared libsyck, it will *not* be used.

    Tag names such as "!!perl/hash:Foo" is blessed into the package "Foo",
    but the "!hs/foo" and "!!hs/Foo" tags are blessed into "hs::Foo". Note
    that this holds true even if the tag contains non-word characters; for
    example, "!haskell.org/Foo" is blessed into "haskell.org::Foo". Please
    use Class::Rebless to cast it into other user-defined packages.

SEE ALSO
    YAML, JSON::Syck

    <http://www.yaml.org/>

AUTHORS
    Audrey Tang <cpan@audreyt.org>

COPYRIGHT
    Copyright 2005-2009 by Audrey Tang <cpan@audreyt.org>.

    This software is released under the MIT license cited below.

    The libsyck code bundled with this library is released by "why the lucky
    stiff", under a BSD-style license. See the COPYING file for details.

  The "MIT" License
    Permission is hereby granted, free of charge, to any person obtaining a
    copy of this software and associated documentation files (the
    "Software"), to deal in the Software without restriction, including
    without limitation the rights to use, copy, modify, merge, publish,
    distribute, sublicense, and/or sell copies of the Software, and to
    permit persons to whom the Software is furnished to do so, subject to
    the following conditions:

    The above copyright notice and this permission notice shall be included
    in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
    OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
    IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
    CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



Changes
[Changes for 1.22_01 - 2012-11-27]

* Fix error message typo: s/existant/existent/g
* Fix for non-printable character detection
* Quote if non-printable characters are present
* Change LoadBlessed to not bless under ANY circumstances

[Changes for 1.21 - 2012-09-20]

* Tests all green on CPAN Testers. releasing to public.

[Changes for 1.21_01 - 2012-06-22]

* Synchronize JSON::Syck with YAML::Syck version number
* Add DumpInto functions (YAML+Syck) which dump into
  a provided scalar instead of a newly allocated one
* Modify DumpFile functions to output directly to the
  specified file/filehandle instead of buffering all
  output in memory.
* Avoid modifying numbers into strings when emitting

[Changes for 1.20 (JSON::Syck 0.43) - 2012-02-14]

* Tests all green on CPAN Testers. releasing to public.

[Changes for 1.20_01 (JSON::Syck 0.43_01) - 2012-02-11]

* Add test to assure META.yml is up to date before release
* Fix for RT 70933 - disable check for empty file handle in LoadFile calls
* Remove call to non-existent code YAML::Syck::_is_openhandle.
* Fix for RT 74785 - compiling with -Werror=format-security fails

[Changes for 1.19 (JSON::Syck 0.42 ) - 2011-11-08]

* Do not run t/leak.t on perls prior to 5.8.9 as 5.8.8
  and Devel::Leak apparently do not mix (perl RT #102606).

[Changes for 1.18 (JSON::Syck 0.41 ) - 2011-11-05]

* Promote to stable release.

[Changes for 1.17_02 (JSON::Syck 0.40_02 ) - 2011-11-03]

* perl RT #102606 - Only lvalue subroutines are supposed to return
  the argument to "return" without copying it. Stop relying on a
  pre-5.15 bug. [Father Chrysostomos]

[Changes for 1.17_01 (JSON::Syck 0.40_01 ) - 2011-02-28]

* Break YAML::Syck dependency on Scalar::Util

[Changes for 1.17 (JSON::Syck 0.40 ) - 2010-11-21]

* Change a C99 comment to a C89 comment

* Upgrade ppport.h from 3.13 to 3.19

* Remove inc/Test/Builder/IO/Scalar.pm from MANIFEST

[Changes for 1.16 (JSON::Syck 0.39 ) - 2010-11-21]

* POD formatting fixes contributed by Justin Hunter

[Changes for 1.15 (JSON::Syck 0.38 ) - 2010-09-23]

* perl RT #78008 - The 5.13.5 development release of perl changed how
  regexes are stringified, breaking our tests. Update the tests to
  deal with the new syntax.

  See fb85c04 (http://perl5.git.perl.org/perl.git/commit/fb85c04) in
  perl.git for context.

[Changes for 1.14 (JSON::Syck 0.37 ) - 2010-08-03]
* RT RT60771 - Further refactor for JSON. if a number is a valid
  perl string, quote it. "1" + "1" = "11" in javascript.
* add YAML tests to make sure HEX and octal preserve their information

[Changes for 1.13 (JSON::Syck 0.36 ) - 2010-08-03]
* RT RT60771 - quoted negative integer and quoted zeroes.
  This was breaking JSON since "0" is true in javascript.

[Changes for 1.12 (JSON::Syck 0.35 ) - 2010-08-03]
* The tie test proved to be very unstable for 5.8 and it's not
  relevant for tie actually. I'm converting it to a TODO for 

[Changes for 1.11 (JSON::Syck 0.35 ) - 2010-08-03]
* Tests are now stable. releasing 1.11

[Changes for 1.10_07 (JSON::Syck 0.34 ) - 2010-07-29]
* META.yml missing

[Changes for 1.10_06 (JSON::Syck 0.34 ) - 2010-07-29]
* Load of in-memory file handles not previously tested (was eval'd 
  out of existance) Corrected code and test to behave correctly.
* Perl 5.8 required to YAML-ify code refs
* Perl 5.6 requires {} to print to a scalar GLOB
* use SvNIOK to test for int or float if we know it's not a string
* Attempt again to get 5.8.7 skipped for tie test

[Changes for 1.10_05 (JSON::Syck 0.34 ) - 2010-07-28]
* Skip failing tie test on perl < 5.8.8 && perl >= 5.7.0x

[Changes for 1.10_04 (JSON::Syck 0.34 ) - 2010-07-19]
* RT 23850 - META.yml of DMAKI/DateTime-Format-Japanese-0.01.tar.gz
  cannot be parsed
* RT 34073 - TODO Tests - Parsing YAML without separator
* RT 20969 - Document die behavior differences with YAML

[Changes for 1.10_03 (JSON::Syck 0.34 ) - 2010-07-19]
* TODO tests to cover noted deficiencies in RT #'s
  26182, 54780, 49404, 42390
* RT 23909 - Loaded self-refererring document is corrupted

[Changes for 1.10_02 (JSON::Syck 0.34 ) - 2010-07-19]
* RT 50227 - Logic bug in deciding what numbers to quote led to some
  interesting corner cases. Now only 9digit integers are supported
  as unquoted. Everything else will be forced to quote
* RT 59458 - convert \%d to \%ld in code to prevent compiler warnings

[Changes for 1.10_01 (JSON::Syck 0.34) - 2010-07-15]
* LoadFile aborts loading an empty file
* RT 50227 - Not quoting strings when data has been sorted numerically
  before being passed to JSON::Syck::Dump
* RT 48327 - use 3 arg form of open() instead of the 2 argument form
* Add TODO tests for RT 52432 - '... X' breaks round trip
* RT 59432 - use lexical file handles for read/write to files
* RT 37236 - Re-enable JSON-basic tests for JSON >= 2.11
* RT 47944 - back quote is a reserved character
* RT 36288 - ability to disable blessing datastructures on load
* Tone down the WARNING section in YAML::Syck's documentation that was
  added in 1.08

[Changes for 1.10 (JSON::Syck 0.33) - 2010-06-06]

1.09 was released with version 0.97 of Module::Install. It had a
critical bug that I reported which broke installations on perl 5.8.3
and older.

This release is equivalent to 1.09 aside from bumping Module::Install
to 0.99 which fixed the bug.

[Changes for 1.09 (JSON::Syck 0.32) - 2010-05-29]

* Exactly equivalent to 1.08_01, aside from bumping the version
  numbers

[Changes for 1.08_01 (JSON::Syck 0.31_01) - 2010-05-23]

* Reset the hash iterator after dumping hashes. Solves RT #54167 and
  the duplicate RT #34166.

* RT #39572: Only print 0.60 compatibility warning if 0.60 or earlier
  is installed.

[Changes for 1.08 (JSON::Syck 0.31) - 2010-05-23]

Like the 1.07_01 test release aside from small documentation
improvements.

* POD link to JSON::XS from JSON::Syck

* Include a WARNING section in the YAML::Syck pod that mentions that
  the module hasn't been actively maintained since 2007, and that

syck_.c
/*
 * syck.c
 *
 * $Author: why $
 * $Date: 2005-01-01 10:06:25 +0800 (.., 01  1 2005) $
 *
 * Copyright (C) 2003 why the lucky stiff
 */
#include <stdio.h>
#include <string.h>

#include "syck.h"

void syck_parser_pop_level( SyckParser * );

/*
 * Custom assert
 */
void 
syck_assert( char *file_name, unsigned line_num )
{
    fflush( NULL );
    fprintf( stderr, "\nAssertion failed: %s, line %u\n",
             file_name, line_num );
    fflush( stderr );
    abort();
}

/*
 * Allocates and copies a string
 */
char *
syck_strndup( char *buf, long len )
{
    char *new = S_ALLOC_N( char, len + 1 );
    S_MEMZERO( new, char, len + 1 );
    S_MEMCPY( new, buf, char, len );
    return new;
}

/*
 * Default FILE IO function
 */
long
syck_io_file_read( char *buf, SyckIoFile *file, long max_size, long skip )
{
    long len = 0;

    ASSERT( file != NULL );

    max_size -= skip;
    len = fread( buf + skip, sizeof( char ), max_size, file->ptr );
    len += skip;
    buf[len] = '\0';

    return len;
}

/*
 * Default string IO function
 */
long
syck_io_str_read( char *buf, SyckIoStr *str, long max_size, long skip )
{
    char *beg;
    long len = 0;

    ASSERT( str != NULL );
    beg = str->ptr;
    if ( max_size >= 0 )
    {
        max_size -= skip;
        if ( max_size <= 0 )  max_size = 0;
        else                  str->ptr += max_size;

        if ( str->ptr > str->end )
        {
            str->ptr = str->end;
        }
    }
    else
    {
        /* Use exact string length */
        while ( str->ptr < str->end ) {
            if (*(str->ptr++) == '\n') break;
        }
    }
    if ( beg < str->ptr )
    {
        len = ( str->ptr - beg );
        S_MEMCPY( buf + skip, beg, char, len );
    }
    len += skip;
    buf[len] = '\0';

    return len;
}

void
syck_parser_reset_levels( SyckParser *p )
{
    while ( p->lvl_idx > 1 )
    {
        syck_parser_pop_level( p );
    }

    if ( p->lvl_idx < 1 )
    {
        p->lvl_idx = 1;
        p->levels[0].spaces = -1;
        p->levels[0].ncount = 0;
        p->levels[0].domain = syck_strndup( "", 0 );
    }
    p->levels[0].status = syck_lvl_header;
}

void
syck_parser_reset_cursor( SyckParser *p )
{
    if ( p->buffer == NULL )
    {
        p->buffer = S_ALLOC_N( char, p->bufsize );
        S_MEMZERO( p->buffer, char, p->bufsize );
    }
    p->buffer[0] = '\0';

    p->cursor = NULL;
    p->lineptr = NULL;
    p->linectptr = NULL;
    p->token = NULL;
    p->toktmp = NULL;
    p->marker = NULL;
    p->limit = NULL;

    p->root = 0;
    p->root_on_error = 0;
    p->linect = 0;
    p->eof = 0;
    p->last_token = 0;
    p->force_token = 0;
}

/*
 * Value to return on a parse error
 */
void
syck_parser_set_root_on_error( SyckParser *p, SYMID roer )
{
    p->root_on_error = roer;
}

/*
 * Allocate the parser
 */
SyckParser *
syck_new_parser()
{
    SyckParser *p;
    p = S_ALLOC( SyckParser );
    S_MEMZERO( p, SyckParser, 1 );
    p->lvl_capa = ALLOC_CT;
    p->levels = S_ALLOC_N( SyckLevel, p->lvl_capa ); 
    p->input_type = syck_yaml_utf8;
    p->io_type = syck_io_str;
    p->io.str = NULL;
    p->syms = NULL;
    p->anchors = NULL;
    p->bad_anchors = NULL;
    p->implicit_typing = 1;
    p->taguri_expansion = 0;
    p->bufsize = SYCK_BUFFERSIZE;
    p->buffer = NULL;
    p->lvl_idx = 0;
    syck_parser_reset_levels( p );
    return p;
}

int
syck_add_sym( SyckParser *p, char *data )
{
    SYMID id = 0;
    if ( p->syms == NULL )
    {
        p->syms = st_init_numtable();
    }
    id = p->syms->num_entries + 1;
    st_insert( p->syms, id, (st_data_t)data );
    return id;
}

int
syck_lookup_sym( SyckParser *p, SYMID id, char **data )
{
    if ( p->syms == NULL ) return 0;
    return st_lookup( p->syms, id, (st_data_t *)data );
}

int
syck_st_free_nodes( char *key, SyckNode *n, char *arg )
{
    if ( n != (void *)1 ) syck_free_node( n );
    n = NULL;
    return ST_CONTINUE;
}

void
syck_st_free( SyckParser *p )
{
    /*
     * Free the anchor tables
     */
    if ( p->anchors != NULL )
    {
        st_foreach( p->anchors, syck_st_free_nodes, 0 );
        st_free_table( p->anchors );
        p->anchors = NULL;
    }

    if ( p->bad_anchors != NULL )
    {
        st_foreach( p->bad_anchors, syck_st_free_nodes, 0 );
        st_free_table( p->bad_anchors );
        p->bad_anchors = NULL;
    }
}

void
syck_free_parser( SyckParser *p )
{
    /*
     * Free the adhoc symbol table
     */
    if ( p->syms != NULL )
    {
        st_free_table( p->syms );
        p->syms = NULL;
    }

    /*
     * Free tables, levels
     */
    syck_st_free( p );
    syck_parser_reset_levels( p );
    S_FREE( p->levels[0].domain );
    S_FREE( p->levels );

    if ( p->buffer != NULL )
    {
        S_FREE( p->buffer );
    }
    free_any_io( p );
    S_FREE( p );
}

void
syck_parser_handler( SyckParser *p, SyckNodeHandler hdlr )
{
    ASSERT( p != NULL );
    p->handler = hdlr;
}

void
syck_parser_implicit_typing( SyckParser *p, int flag )
{
    p->implicit_typing = ( flag == 0 ? 0 : 1 );
}

void
syck_parser_taguri_expansion( SyckParser *p, int flag )
{
    p->taguri_expansion = ( flag == 0 ? 0 : 1 );
}

void
syck_parser_error_handler( SyckParser *p, SyckErrorHandler hdlr )
{
    ASSERT( p != NULL );
    p->error_handler = hdlr;
}

void
syck_parser_bad_anchor_handler( SyckParser *p, SyckBadAnchorHandler hdlr )
{
    ASSERT( p != NULL );
    p->bad_anchor_handler = hdlr;
}

void
syck_parser_set_input_type( SyckParser *p, enum syck_parser_input input_type )
{
    ASSERT( p != NULL );
    p->input_type = input_type;
}

void
syck_parser_file( SyckParser *p, FILE *fp, SyckIoFileRead read )
{
    ASSERT( p != NULL );
    free_any_io( p );
	syck_parser_reset_cursor( p );
    p->io_type = syck_io_file;
    p->io.file = S_ALLOC( SyckIoFile );
    p->io.file->ptr = fp;
    if ( read != NULL )
    {
        p->io.file->read = read;
    }
    else
    {
        p->io.file->read = syck_io_file_read;
    }
}

void
syck_parser_str( SyckParser *p, char *ptr, long len, SyckIoStrRead read )
{
    ASSERT( p != NULL );
    free_any_io( p );
	syck_parser_reset_cursor( p );
    p->io_type = syck_io_str;
    p->io.str = S_ALLOC( SyckIoStr );
    p->io.str->beg = ptr;
    p->io.str->ptr = ptr;
    p->io.str->end = ptr + len;
    if ( read != NULL )
    {
        p->io.str->read = read;
    }
    else
    {
        p->io.str->read = syck_io_str_read;
    }
}

void
syck_parser_str_auto( SyckParser *p, char *ptr, SyckIoStrRead read )
{
    syck_parser_str( p, ptr, strlen( ptr ), read );
}

SyckLevel *
syck_parser_current_level( SyckParser *p )
{
    return &p->levels[p->lvl_idx-1];
}

void
syck_parser_pop_level( SyckParser *p )
{
    ASSERT( p != NULL );

    /* The root level should never be popped */
    if ( p->lvl_idx <= 1 ) return;

    p->lvl_idx -= 1;
    free( p->levels[p->lvl_idx].domain );
}

void 
syck_parser_add_level( SyckParser *p, int len, enum syck_level_status status )
{
    ASSERT( p != NULL );
    if ( p->lvl_idx + 1 > p->lvl_capa )
    {
        p->lvl_capa += ALLOC_CT;
        S_REALLOC_N( p->levels, SyckLevel, p->lvl_capa );
    }

    ASSERT( len > p->levels[p->lvl_idx-1].spaces );
    p->levels[p->lvl_idx].spaces = len;
    p->levels[p->lvl_idx].ncount = 0;
    p->levels[p->lvl_idx].domain = syck_strndup( p->levels[p->lvl_idx-1].domain, strlen( p->levels[p->lvl_idx-1].domain ) );
    p->levels[p->lvl_idx].status = status;
    p->lvl_idx += 1;
}

void
free_any_io( SyckParser *p )
{
    ASSERT( p != NULL );
    switch ( p->io_type )
    {
        case syck_io_str:
            if ( p->io.str != NULL ) 
            {
                S_FREE( p->io.str );
                p->io.str = NULL;
            }
        break;

        case syck_io_file:
            if ( p->io.file != NULL ) 
            {
                S_FREE( p->io.file );
                p->io.file = NULL;
            }
        break;
    }
}

long
syck_move_tokens( SyckParser *p )
{
    long count, skip;
    ASSERT( p->buffer != NULL );

    if ( p->token == NULL )
        return 0;

    skip = p->limit - p->token;
    if ( skip < 0 )
        return 0;

    if ( ( count = p->token - p->buffer ) )
    {
        S_MEMMOVE( p->buffer, p->token, char, skip );
        p->token = p->buffer;
        p->marker -= count;
        p->cursor -= count;
        p->toktmp -= count;
        p->limit -= count;
        p->lineptr -= count;
        p->linectptr -= count;
    }
    return skip;
}

void
syck_check_limit( SyckParser *p, long len )
{
    if ( p->cursor == NULL )
    {
        p->cursor = p->buffer;
        p->lineptr = p->buffer;
        p->linectptr = p->buffer;
        p->marker = p->buffer;
    }
    p->limit = p->buffer + len;
}

long
syck_parser_read( SyckParser *p )
{
    long len = 0;
    long skip = 0;
    ASSERT( p != NULL );
    switch ( p->io_type )
    {
        case syck_io_str:
            skip = syck_move_tokens( p );
            len = (p->io.str->read)( p->buffer, p->io.str, SYCK_BUFFERSIZE - 1, skip );
            break;

        case syck_io_file:
            skip = syck_move_tokens( p );
            len = (p->io.file->read)( p->buffer, p->io.file, SYCK_BUFFERSIZE - 1, skip );
            break;
    }
    syck_check_limit( p, len );
    return len;
}

long
syck_parser_readlen( SyckParser *p, long max_size )
{
    long len = 0;
    long skip = 0;
    ASSERT( p != NULL );
    switch ( p->io_type )
    {
        case syck_io_str:
            skip = syck_move_tokens( p );
            len = (p->io.str->read)( p->buffer, p->io.str, max_size, skip );
            break;

        case syck_io_file:
            skip = syck_move_tokens( p );
            len = (p->io.file->read)( p->buffer, p->io.file, max_size, skip );
            break;
    }
    syck_check_limit( p, len );
    return len;
}

SYMID
syck_parse( SyckParser *p )
{
    ASSERT( p != NULL );

    syck_st_free( p );
    syck_parser_reset_levels( p );
    syckparse( p );
    return p->root;
}

void
syck_default_error_handler( SyckParser *p, char *msg )
{
    printf( "Error at [Line %d, Col %ld]: %s\n",
        p->linect,
        p->cursor - p->lineptr,
        msg );
}

int syck_str_is_unquotable_integer(char* str, long len) {
	int idx;

	if(!str) return 0; /* Don't parse null strings */
	if(len < 1) return 0; /* empty strings can't be numbers */
	if(len > 9) return 0; /* Ints larger than 9 digits (32bit) might not portable. Force a string. */

	if(str[0] == '-' && (len < 2 || str[1] > '9' || str[1] < '1')) return 0;
	if(str[0] > '9' || str[0] < '1') return 0;

	/* Look for illegal characters */
	for ( idx = 1; idx < len; idx++ ) {
		if(!isdigit(str[idx])) return 0;
	}

	return 1;
}


lib/YAML/Syck.pm
package YAML::Syck;
# See documentation after the __END__ mark.

use strict;
use vars qw(
    @ISA @EXPORT $VERSION
    $Headless $SortKeys $SingleQuote
    $ImplicitBinary $ImplicitTyping $ImplicitUnicode 
    $UseCode $LoadCode $DumpCode
    $DeparseObject
);
use 5.00307;
use Exporter;

BEGIN {
    $VERSION = '1.06';
    @EXPORT  = qw( Dump Load DumpFile LoadFile );
    @ISA     = qw( Exporter );

    $SortKeys = 1;

    local $@;
    eval {
        require XSLoader;
        XSLoader::load(__PACKAGE__, $VERSION);
        1;
    } or do {
        require DynaLoader;
        push @ISA, 'DynaLoader';
        __PACKAGE__->bootstrap($VERSION);
    };

}

use constant QR_MAP => {
    ''   => sub { qr{$_[0]}     }, 
    x    => sub { qr{$_[0]}x    }, 
    i    => sub { qr{$_[0]}i    }, 
    s    => sub { qr{$_[0]}s    }, 
    m    => sub { qr{$_[0]}m    }, 
    ix   => sub { qr{$_[0]}ix   }, 
    sx   => sub { qr{$_[0]}sx   }, 
    mx   => sub { qr{$_[0]}mx   }, 
    si   => sub { qr{$_[0]}si   }, 
    mi   => sub { qr{$_[0]}mi   }, 
    ms   => sub { qr{$_[0]}sm   }, 
    six  => sub { qr{$_[0]}six  }, 
    mix  => sub { qr{$_[0]}mix  }, 
    msx  => sub { qr{$_[0]}msx  }, 
    msi  => sub { qr{$_[0]}msi  }, 
    msix => sub { qr{$_[0]}msix }, 
};

sub __qr_helper {
    if ($_[0] =~ /\A  \(\?  ([ixsm]*)  (?:-  (?:[ixsm]*))?  : (.*) \)  \z/x) {
        my $sub = QR_MAP()->{$1} || QR_MAP()->{''};
        &$sub($2);
    }
    else {
        qr/$_[0]/;
    }
}

sub Dump {
    $#_ ? join('', map { YAML::Syck::DumpYAML($_) } @_)
        : YAML::Syck::DumpYAML($_[0]);
}

sub Load {
    if (wantarray) {
        my ($rv) = YAML::Syck::LoadYAML($_[0]);
        @{$rv};
    }
    else {
        YAML::Syck::LoadYAML($_[0]);
    }
}

# NOTE. The code below (_is_openhandle) avoids to require/load
# Scalar::Util unless it is given a ref or glob
# as an argument. That is purposeful, so to avoid
# the need for this dependency unless strictly necessary.
# If that was not the case, Scalar::Util::openhandle could
# be used directly.

sub _is_openhandle {
    my $h = shift;
    if ( ref($h) || ref(\$h) eq 'GLOB' ) {
        require Scalar::Util;
        return Scalar::Util::openhandle($h);
    } else {
        return undef;
    }
}

sub DumpFile {
    my $file = shift;
    if ( _is_openhandle($file) ) {
        if ($#_) {
            print {$file} YAML::Syck::DumpYAML($_) for @_;
        }
        else {
            print {$file} YAML::Syck::DumpYAML($_[0]);
        }
    }
    else {
        local *FH;
        open FH, "> $file" or die "Cannot write to $file: $!";
        if ($#_) {
            print FH YAML::Syck::DumpYAML($_) for @_;
        }
        else {
            print FH YAML::Syck::DumpYAML($_[0]);
        }
        close FH;
    }
}

sub LoadFile {
    my $file = shift;
    if ( _is_openhandle($file) ) {
        Load(do { local $/; <$file> });
    }
    else {
        local *FH;
        open FH, "< $file" or die "Cannot read from $file: $!";
        Load(do { local $/; <FH> });
    }
}

1;

__END__
=pod

=head1 NAME 

YAML::Syck - Fast, lightweight YAML loader and dumper

=head1 VERSION

This document describes version 1.06 of YAML::Syck, released April 25, 2009.

=head1 SYNOPSIS

    use YAML::Syck;

    # Set this for interoperability with other YAML/Syck bindings:
    # e.g. Load('Yes') becomes 1 and Load('No') becomes ''.
    $YAML::Syck::ImplicitTyping = 1;

    $data = Load($yaml);
    $yaml = Dump($data);

    # $file can be an IO object, or a filename
    $data = LoadFile($file);
    DumpFile($file, $data);

    # A string with multiple YAML streams in it
    $yaml = Dump(@data);
    @data = Load($yaml);

=head1 DESCRIPTION

This module provides a Perl interface to the B<libsyck> data serialization
library.  It exports the C<Dump> and C<Load> functions for converting
Perl data structures to YAML strings, and the other way around.

B<NOTE>: If you are working with other language's YAML/Syck bindings
(such as Ruby), please set C<$YAML::Syck::ImplicitTyping> to C<1> before
calling the C<Load>/C<Dump> functions.  The default setting is for
preserving backward-compatibility with C<YAML.pm>.

=head1 FLAGS

=head2 $YAML::Syck::Headless

Defaults to false.  Setting this to a true value will make C<Dump> omit the
leading C<---\n> marker.

=head2 $YAML::Syck::SortKeys

Defaults to false.  Setting this to a true value will make C<Dump> sort
hash keys.

=head2 $YAML::Syck::SingleQuote

Defaults to false.  Setting this to a true value will make C<Dump> always emit
single quotes instead of bare strings.

=head2 $YAML::Syck::ImplicitTyping

Defaults to false.  Setting this to a true value will make C<Load> recognize
various implicit types in YAML, such as unquoted C<true>, C<false>, as well as
integers and floating-point numbers.  Otherwise, only C<~> is recognized to
be C<undef>.

=head2 $YAML::Syck::ImplicitUnicode

Defaults to false.  For Perl 5.8.0 or later, setting this to a true value will
make C<Load> set Unicode flag on for every string that contains valid UTF8
sequences, and make C<Dump> return a unicode string.

Regardless of this flag, Unicode strings are dumped verbatim without escaping;
byte strings with high-bit set will be dumped with backslash escaping.

However, because YAML does not distinguish between these two kinds of strings,
so this flag will affect loading of both variants of strings.

=head2 $YAML::Syck::ImplicitBinary

Defaults to false.  For Perl 5.8.0 or later, setting this to a true value will
make C<Dump> generate Base64-encoded C<!!binary> data for all non-Unicode
scalars containing high-bit bytes.

=head2 $YAML::Syck::UseCode / $YAML::Syck::LoadCode / $YAML::Syck::DumpCode

These flags control whether or not to try and eval/deparse perl source code;
each of them defaults to false.

Setting C<$YAML::Syck::UseCode> to a true value is equivalent to setting
both C<$YAML::Syck::LoadCode> and C<$YAML::Syck::DumpCode> to true.

=head1 BUGS

Dumping Glob/IO values does not work yet.

=head1 CAVEATS

This module implements the YAML 1.0 spec.  To deal with data in YAML 1.1, 
please use the C<YAML::XS> module instead.

The current implementation bundles libsyck source code; if your system has a
site-wide shared libsyck, it will I<not> be used.

Tag names such as C<!!perl/hash:Foo> is blessed into the package C<Foo>, but
the C<!hs/foo> and C<!!hs/Foo> tags are blessed into C<hs::Foo>.  Note that
this holds true even if the tag contains non-word characters; for example,
C<!haskell.org/Foo> is blessed into C<haskell.org::Foo>.  Please use
L<Class::Rebless> to cast it into other user-defined packages.

=head1 SEE ALSO

L<YAML>, L<JSON::Syck>

L<http://www.yaml.org/>

=head1 AUTHORS

Audrey Tang E<lt>cpan@audreyt.orgE<gt>

=head1 COPYRIGHT

Copyright 2005-2009 by Audrey Tang E<lt>cpan@audreyt.orgE<gt>.

This software is released under the MIT license cited below.

The F<libsyck> code bundled with this library is released by
"why the lucky stiff", under a BSD-style license.  See the F<COPYING>
file for details.

=head2 The "MIT" License

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.

=cut
